lib/fluent/plugin/out_bigquery.rb

```diff
@@ -382,115 +311,177 @@ def format(tag, time, record)
       end
 
-      def _write(chunk, table_format, template_suffix_format)
-        rows = []
-        chunk.msgpack_each do |row_object|
-          # TODO: row size limit
-          rows << row_object.deep_symbolize_keys
+      def write(chunk)
+        table_id_format = @tables_mutex.synchronize do
+          t = @tables_queue.shift
+          @tables_queue.push t
+          t
         end
+        template_suffix_format = @template_suffix
+        _write(chunk, table_id_format, template_suffix_format)
+      end
 
-        now = Time.at(Fluent::Engine.now)
-        group = rows.group_by do |row|
-          [
-            generate_table_id(table_format, now, row, chunk),
-            template_suffix_format ? generate_table_id(template_suffix_format, now, row, chunk) : nil,
-          ]
-        end
-        group.each do |(table_id, template_suffix), group_rows|
-          insert(table_id, group_rows, template_suffix)
-        end
+      # for delayed commit
+      def try_write(chunk)
+        write(chunk)
+        commit_write(chunk.unique_id, delayed: true)
       end
 
-      def insert(table_id, rows, template_suffix)
-        writer.insert_rows(@project, @dataset, table_id, rows, skip_invalid_rows: @skip_invalid_rows, ignore_unknown_values: @ignore_unknown_values, template_suffix: template_suffix)
-      rescue Fluent::BigQuery::Writer::Error => e
-        if @auto_create_table && e.status_code == 404 && /Not Found: Table/i =~ e.message
-          # Table Not Found: Auto Create Table
-          writer.create_table(@project, @dataset, table_id, @fields)
-          raise "table created. send rows next time."
-        end
+      def fetch_schema(allow_overwrite = true)
+        table_id = nil
+        @fetch_schema_mutex.synchronize do
+          if Fluent::Engine.now - @last_fetch_schema_time > @schema_cache_expire
+            table_id = @fetch_schema_table || @tablelist[0]
+            schema = writer.fetch_schema(@project, @dataset, table_id)
+
+            if schema
+              if allow_overwrite
+                fields = Fluent::BigQuery::RecordSchema.new("record")
+                fields.load_schema(schema, allow_overwrite)
+                @fields = fields
+              else
+                @fields.load_schema(schema, allow_overwrite)
+              end
+            else
+              if @fields.empty?
+                raise "failed to fetch schema from bigquery"
+              else
+                log.warn "#{table_id} uses previous schema"
+              end
+            end
 
-        if e.retryable?
-          raise e # TODO: error class
-        elsif @secondary
-          flush_secondary(@secondary)
+            @last_fetch_schema_time = Fluent::Engine.now
+          end
         end
       end
-    end
 
-    module LoadImplementation
-      def format(tag, time, record)
-        fetch_schema if @fetch_schema_table
+      module InsertImplementation
+        def _write(chunk, table_format, template_suffix_format)
```

Comment:
To avoid such cases, enforce users to control buffer chunk keys to split records to be inserted separately, into different chunks.
Check placeholders for `table_format`, and raise error if it's inconsistent. An example is here: https://github.com/fluent/fluentd/pull/1257/files#diff-453ef4fa0f6e84a0194b19878f4944d3R118

If a chunk has only records for a `table_id`, `write` method calls just one `insert`, and retries will be called correctly just for failed records.